### PR TITLE
Add udp2tcp bindings IOS-218

### PR DESCRIPTION
**Pull request by pronebird**
This PR adds Rust bindings for `udp2tcp` and integration test that simulates TCP endpoint and runs traffic over udp <-> udp2tcp <-> tcp server mock.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853)

### CODE REVIEW

File: ios/TunnelObfuscation/tunnel-obfuscator-proxy/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+[target.'cfg(target_os = "ios")'.dependencies]
+tunnel-obfuscation = { path = "../../../tunnel-obfuscation" }
+tokio = { version = "1", features = ["sync"] }
+libc = "0.2"
+log = "0.4"
+oslog = "0.2"
+
+[target.'cfg(target_os = "ios")'.build-dependencies]
+cbindgen = "0.24"
```
Comment (faern) [inline ios/TunnelObfuscation/tunnel-obfuscator-proxy/Cargo.toml:20]: Please don't add this with default features. See #4850

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853)* status: 0 of 18 files reviewed, 1 unresolved discussion (waiting on @faern)

___
*[`ios/TunnelObfuscation/tunnel-obfuscator-proxy/Cargo.toml` line 20 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853#gh-1246602220:-NZA-TXx4PZnUsyW_sHI:bt2vr1b) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6e40cb14f7b83a873086800018e5b627187432dc/ios/TunnelObfuscation/tunnel-obfuscator-proxy/Cargo.toml#L20)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Please don't add this with default features. See #4850
</blockquote></details>

Fixed.

**Review by faern — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853)* status: 0 of 18 files reviewed, all discussions resolved

___
*[`ios/TunnelObfuscation/tunnel-obfuscator-proxy/Cargo.toml` line 20 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853#gh-1246602220:-NZPuuVUC1fZeX6YfKNC:bzfotig) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6e40cb14f7b83a873086800018e5b627187432dc/ios/TunnelObfuscation/tunnel-obfuscator-proxy/Cargo.toml#L20)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

Thanks! In the other PR now I ban `clap 3` from our dependency tree to avoid it accidentally being added later again.

**Review by buggmagnet — COMMENTED** [review]
> Reviewed 7 of 18 files at r1, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853)* status: 6 of 19 files reviewed, 2 unresolved discussions (waiting on @pronebird)

___
*[`ios/TunnelObfuscation/tunnel-obfuscator-proxy/build.rs` line 1 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853#-NZf4jVc2NXsja3hsYMH:-NZf4jVc2NXsja3hsYMI:bqwdd9t) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9b78d28af9af63d1f638fdc0d007e79280e19ca5/ios/TunnelObfuscation/tunnel-obfuscator-proxy/build.rs#L1)):*
> ```XML
> #[cfg(target_os = "ios")]
> ```

I wonder if we want to enable this on macOS as well to help develop the FFI API while we're working on it
___
*[`ios/TunnelObfuscation/tunnel-obfuscator-proxy/src/ffi.rs` line 10 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853#-NZf4t9P-7kP_2zTNh-6:-NZf4t9P-7kP_2zTNh-7:bk4smqq) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9b78d28af9af63d1f638fdc0d007e79280e19ca5/ios/TunnelObfuscation/tunnel-obfuscator-proxy/src/ffi.rs#L10)):*
> ```XML
> 
> #[repr(C)]
> pub struct ProxyHandle {
> ```

This has the same name as the `ProxyHandle` struct used by the `Shadowsocks` crate, it's okay because they are the same structs at the moment, but we might want to either rename one of them, or unify the implementation in the future. 

Thoughts @pinkisemils  @faern  ?

**Review by pinkisemils — COMMENTED** [review]
> Reviewed 1 of 18 files at r1, 1 of 4 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853)* status: 6 of 19 files reviewed, 3 unresolved discussions (waiting on @buggmagnet, @faern, and @pronebird)

___
*[`ios/TunnelObfuscation/tunnel-obfuscator-proxy/src/ffi.rs` line 10 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853#-NZf4t9P-7kP_2zTNh-6:-NZf7-Uo742zzxAloGsA:b-9twi7w) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9b78d28af9af63d1f638fdc0d007e79280e19ca5/ios/TunnelObfuscation/tunnel-obfuscator-proxy/src/ffi.rs#L10)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

This has the same name as the `ProxyHandle` struct used by the `Shadowsocks` crate, it's okay because they are the same structs at the moment, but we might want to either rename one of them, or unify the implementation in the future. 

Thoughts @pinkisemils  @faern  ?
</blockquote></details>

I think this should be fixed. Both should be prefixed with a better name. The shadowsocks handle should indicate it's a shadowsocks proxy for a TCP endpoint, this one should be called udp2tcp proxy handle.
___
*[`ios/TunnelObfuscation/tunnel-obfuscator-proxy/src/ffi.rs` line 23 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4853#-NZf7quH5yrlQAi1l-IM:-NZf7quH5yrlQAi1l-IN:b2gkrnx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9b78d28af9af63d1f638fdc0d007e79280e19ca5/ios/TunnelObfuscation/tunnel-obfuscator-proxy/src/ffi.rs#L23)):*
> ```XML
> ) -> i32 {
>     INIT_LOGGING.call_once(|| {
>         let _ = oslog::OsLogger::new("net.mullvad.MullvadVPN.TunnelObfuscatorProxy")
> ```

I wonder if we don't end up trashing the logging here, since this might also initialize `OsLog`. 

We could improve this and just pass a logger from Swift to Rust instead of trying to use OsLog _directly_. But this is unrelated to this particular PR.

**Pull request closed without merging**